### PR TITLE
feat: decrease tick spacing after pool creation with gov prop

### PR DESCRIPTION
**Pull request by czarcas7ic**
Closes: #4938 

## What is the purpose of the change

We currently have no way to change a pools tick spacing after pool creation. Also, if there was a way, we would need it to be triggered by governance. This PR both introduces this logic as well as the respective gov method and wiring.

## Brief Changelog

- Adds TickSpacingDecreaseProposal gov proto
- Introduces gov handler to concentrated liquidity
  - Can be expanded with other gov messages in the future
- Adds a setter method for tick spacing
- Adds DecreaseConcentratedPoolTickSpacing keeper method that ensures the given tick spacing is less than the current tick spacing,  and is one of the authorized tick spacings

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (yes / no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (yes / no)
  - How is the feature or change documented? (not applicable   /   specification (`x/<module>/spec/`)  /  [Osmosis docs repo](https://github.com/osmosis-labs/docs)   /   not documented)

### CODE REVIEW

File: x/concentrated-liquidity/pool.go [hunk 1]
```diff
@@ -253,6 +253,29 @@ func (k Keeper) GetSerializedPools(ctx sdk.Context, pagination *query.PageReques
 	return anys, pageRes, err
 }
 
+// DecreaseConcentratedPoolTickSpacing decreases the tick spacing of the given pools to the given tick spacings.
+// It returns an error if the tick spacing is not one of the authorized tick spacings or is not less than the current tick spacing of the respective pool.
+func (k Keeper) DecreaseConcentratedPoolTickSpacing(ctx sdk.Context, poolIdToTickSpacingRecord []types.PoolIdToTickSpacingRecord) error {
```
Comment (AlpinYukseloglu) [inline x/concentrated-liquidity/pool.go:258]: Just to clarify, decreasing tick spacing is equivalent to increasing the number of valid ticks right? Ideally we can make this explicit in the godoc

Comment (czarcas7ic) [reply x/concentrated-liquidity/pool.go:259]: added a clarifying comment here https://github.com/osmosis-labs/osmosis/pull/4982/commits/87cfdd1928c4edb67bec6393fe1d10422a096958

**Review by p0mvn — APPROVED** [review]
This LGTM

File: app/keepers/keepers.go [hunk 2]
```diff
@@ -473,7 +473,8 @@ func (appKeepers *AppKeepers) InitNormalKeepers(
 		AddRoute(txfeestypes.RouterKey, txfees.NewUpdateFeeTokenProposalHandler(*appKeepers.TxFeesKeeper)).
 		AddRoute(superfluidtypes.RouterKey, superfluid.NewSuperfluidProposalHandler(*appKeepers.SuperfluidKeeper, *appKeepers.EpochsKeeper, *appKeepers.GAMMKeeper)).
 		AddRoute(protorevtypes.RouterKey, protorev.NewProtoRevProposalHandler(*appKeepers.ProtoRevKeeper)).
-		AddRoute(gammtypes.RouterKey, gamm.NewMigrationRecordHandler(*appKeepers.GAMMKeeper))
+		AddRoute(gammtypes.RouterKey, gamm.NewMigrationRecordHandler(*appKeepers.GAMMKeeper)).
```
Comment (mattverse) [inline app/keepers/keepers.go:476]: lol rip with the other cl gov prop PR 😢

File: app/keepers/keepers.go [hunk 3]
```diff
@@ -473,7 +473,8 @@ func (appKeepers *AppKeepers) InitNormalKeepers(
...
+		AddRoute(concentratedliquiditytypes.RouterKey, concentratedliquidity.NewConcentratedLiquidityProposalHandler(*appKeepers.ConcentratedLiquidityKeeper))
```
Comment (mattverse) [inline app/keepers/keepers.go:477]: Do we also not need to add the client handler as in https://github.com/osmosis-labs/osmosis/pull/4976/files#diff-95b3d541ce5c3238d75044a49770ab16c1bf1913ed0064a14897ed8474eba1b7R82?

**Review by mattverse — COMMENTED** [review]
Nice work! Following questions:
1. Do we not need to register client for the gov proposal?
2. Do we not need to register codec for the new gov proposal?
3. Can we add e2e testing for this gov propoals?

File: app/keepers/keepers.go (hunk 3 above)
Comment (czarcas7ic) [reply app/keepers/keepers.go:477]: There are so many steps to adding a new message -__-, thanks for catching these.

Added codec, client, and cli here https://github.com/osmosis-labs/osmosis/pull/4982/commits/ae766e8ce75efba7721ffc658c3da9a4b484d2ae

File: x/concentrated-liquidity/types/gov.go [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+// String returns a string containing the decrease tick spacing proposal.
+func (p TickSpacingDecreaseProposal) String() string {
+	recordsStr := ""
+	for _, record := range p.PoolIdToTickSpacingRecords {
+		recordsStr = recordsStr + fmt.Sprintf("(PoolID: %d, NewTickSpacing: %d) ", record.PoolId, record.NewTickSpacing)
+	}
+
+	var b strings.Builder
+	b.WriteString(fmt.Sprintf(`Update Pools Tick Spacing Proposal:
```
Comment (mattverse) [inline x/concentrated-liquidity/types/gov.go:67]:
```suggestion
	b.WriteString(fmt.Sprintf(`Decrease Pools Tick Spacing Proposal:
```

File: x/concentrated-liquidity/client/cli/tx.go [hunk 5]
```diff
@@ -76,3 +86,115 @@ func NewCreateIncentiveCmd() (*osmocli.TxCliDesc, *types.MsgCreateIncentive) {
...
+		Description:                description,
+		PoolIdToTickSpacingRecords: poolIdToTickSpacingRecords,
+	}
+	return content, nil
+}
+
+func parsePoolIdToTickSpacingRecords(cmd *cobra.Command) ([]types.PoolIdToTickSpacingRecord, error) {
+	assetsStr, err := cmd.Flags().GetString(FlagPoolIdToTickSpacingRecords)
+	if err != nil {
+		return nil, err
```
Comment (mattverse) [inline x/concentrated-liquidity/client/cli/tx.go:173]: can we add a check to see if the len(assets) is even number?

**Review by mattverse — APPROVED** [review]
LGTM

File: x/concentrated-liquidity/client/cli/tx.go (hunk 5 above)
Comment (czarcas7ic) [reply x/concentrated-liquidity/client/cli/tx.go:174]: Makes sense to me, added here https://github.com/osmosis-labs/osmosis/pull/4982/commits/54103588e1ad7c771a7b6532ca9bebfea7eca607

**Pull request merged**